### PR TITLE
Added function "removeGarbageParametersFromURL()"

### LINKED ISSUE #6498: Add `removeGarbageParametersFromURL` function
**Use case**
Remove long base64 and hex encoded paramters that look random. Example:

`http://yandex.ru/clck/jsredir?from=yandex.ru;search%2F;web;;&text=&etext=1004&cst=AiuY0DBWFJ5Hyx_fyvalFPA3abBqdnSOApSiLPWwVkIeiz46AroRCQXrfJ8M5oYTWorWEWccK4Kw_QEhSDD6X4nGMT4OabEk0xnry4NtnOEzWFPU4iTzunSVVjkuY7CYolnD7hb04cMRv7iMnaO8LjNm0hxqvwN9sXCzUYeXp_muLsdY4W99_U5MJKGmz7IAmR5-ceoAoaBB2XGYAS9BTYKbbvlmneBpbf_SwAd_6OOACXtLmRXqXad3AQbcArYE8LCO0zmE9vpha3yoT0jl8pd9CUmbGZR5nA3sf5TcDFTpr5nYaOdxjmHep2cZeW3QHvPtKA2xWXW6qzGrQeZ1SEOPcJ1afJqmAHisup90hhNYyl2hxl8xn_DtCRbJqYHb88JtuQ3591EGW42wPZhSbxBFdU0KIZN3c_VZOmk6avzKzqG_kJpjPObWXbh9qs0S23WxDGCcPUrIzi3ESSLv1qgaRhqkfjBc57BFVA4RxlljpKQdeVeTbklJgqptznf1aHZQ2wYARBzC_jvv994MCTZIus_NctCMWoSaU74OaMmo0h5ScYLI2CWy6nj5PbhCrgeLsaEBVOQT9xoLSoCRfJ78xI_T1ruuD3QBJmHY6YW8f5UM36LRbzhd5vmNTPRvrs2wcCFhF_w&l10n=ru&cts=1458904227333&mc=1.584962500721156`

Note: these parameters are usually not "garbage" but they are impossible to decrypt and they will waste space on disk. We are already removing these parameters from Yandex, Google, etc. before inserting into ClickHouse.

**Implementation**
For parameters of some minimal length, quickly calculate some statistics like:
- does not contain symbols that are not in base64 or base64-for-URL;
- distribution looks uniform.

### PR DESCRIPTION
**Pull request by aterehov**
Fixes: #6498

Added `removeGarbageParameterFromURL()` function that removes parameters of some minimal length that look random.

Added functions interface that allows creating functions accepting string and unsigned integer and returning string

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a function that removes randomly generated parameters from URL

### CODE REVIEW

**Review by rschu1ze — COMMENTED** [review]
@aterehov Thanks for this contribution. This PR would need
- a test in tests/queries/0_stateless (see [here](https://clickhouse.com/docs/en/development/tests))
- documentation [here](https://clickhouse.com/docs/en/sql-reference/functions/url-functions)

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 4f2b452d6992cbb3d11ce714283a5668cccd8623 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49305/4f2b452d6992cbb3d11ce714283a5668cccd8623/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+        /// Set length
+        res_offsets[0] = 0;
+
+        /// Prepare pointers
+        size_t prev_offset = 0;
+        size_t cur_offset = offsets[0];
+        const char * pdata = reinterpret_cast<const char *>(&data[prev_offset]);
+        const char * pdata_end = reinterpret_cast<const char *>(&data[cur_offset]) - 1;
+        size_t size = pdata_end - pdata;
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:314]: `url_length`

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+        const char * read = pdata;
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:315]: `cur` is (as far as I see) the most common term for the current read pointer

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+
+        /// Calculate length of webpage address and copy it
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:317]: Maybe `Calculate length of the protocol, domain and path`

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        int64_t url_size = -1;
+        for (size_t i = 0; i < size; ++i)
+        {
+            if (read[i] == '?')
+            {
+                url_size = i;
+                break;
+            }
+            else if (read[i] == 0)
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:326]: Which situation does this handle? Is `0` in an URL valid?

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+        }
+        return process_content(c_start, p_end, min_len);
+    }
+
+    /// @brief Add data to string
+    /// @param res_data String to add data to
+    /// @param size length of data to be added
+    /// @param pdata pointer to the first character of data
+    /// @param res_offsets offsets of output string
+    static void copy(ColumnString::Chars & res_data, size_t size, const char * pdata, ColumnString::Offsets & res_offsets)
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:270]: `copyIntoOutput`.

Also let's bring the parameters into a more logical order: `const char * from, size_t from_length, ColumnString::Chars & res_data, ColumnString::Offsets & res_offsets`

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+        size_t old_size = res_data.size();
+        res_data.resize(old_size + size);
+        memcpySmallAllowReadWriteOverflow15(&res_data[old_size], pdata, size);
+        res_offsets[0] += size;
+    }
+
+    /// @brief End the string with null terminator
+    /// @param res_data String to finalize
+    /// @param res_offsets Offsets of string
+    static void finalize_data(ColumnString::Chars & res_data, ColumnString::Offsets & res_offsets)
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:281]: camelCase: `finalizeData`

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+            *write = *i + correction;
+            ++write;
+        }
+
+        bool ret = pearsonIsUniform(n_c_start, n_c_end, '0', '?');
+        free(n_c_start);
+        return !ret;
+    }
+
+    /// @brief Check if string is long enough and if it could be hex or base64 for url
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:199]: Doxygen-style parameter docs are quite rare in the codebase. I would say if the parameters have expressive enough names then we can remove it and just keep l. 199.

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+            if (check_base64
+                && !(
+                    ('A' <= *i && *i <= 'Z') || ('a' <= *i && *i <= 'z') || ('0' <= *i && *i <= '9') || (*i == '-') || (*i == '_')
+                    || (*i == '=')))
+            {
+                check_base64 = false;
+                break;
+            }
+        }
+        if (!check_hex && !check_base64)
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:228]: (cosmetic: let's omit {} in single-line if/for/while statements)

File: src/Functions/URL/removeGarbageParametersFromURL.cpp [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+    /// @param c_start Pointer to the first symbol of the string
+    /// @param c_end Pointer to the symbol following the last symbol of the string
+    /// @param a First character of uniform distribution
+    /// @param b Last character of uniform distribution
+    /// @return true - if there is 95%+ probability the string matches uniform distribution with given parameters, false - otherwise
+    static bool pearsonIsUniform(const char * c_start, const char * c_end, char a, char b)
+    {
+        /// Deciding how much intervals to split symbols into and whether the string is long enough to be handled
+        int64_t length = c_end - c_start;
+        int l = static_cast<int>(3.32 * log10(length)) + 1;
```
Comment (rschu1ze) [inline src/Functions/URL/removeGarbageParametersFromURL.cpp:27]: To a random future reader, this looks like too much magic. Could you please document the kind of test/algorithm used in `pearsonIsUniform()?

**Comment by rschu1ze** [comment]
@aterehov Thanks for updating this PR and your effort! This seems a student/coursework project if I understand correctly.

To be honest, I am hesitant to merge this PR. Not because it has bad quality but because I don't really see practical use cases for the new functionality. 1. Do random URL parameters really occur in practice? It is very unlikely that parameters were garbled during transmission. I can only imagine that a malicious attacker constructs garbage or an accidental user input error. 2. And if such parameters occur, does the result of a magic repair function (which can and does only work probabilistically) produce a URL that still bears useful semantics for the code / program interpreting this URL? There is a principle [garbage in, garbage out](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out) which in my view also applies to URLs. Any attempt to repair them is basically doomed to fail in the general case.

**Comment by aterehov** [comment]
@rschu1ze Yes, this used to be a coursework project when I started doing it. Later I decided to update the function so it could be merged into ClickHouse. 

The function seems to be finished now, and the errors in checks do not seem to be related to removeGarbageParametersFromURL. You can continue working on it and merge it into ClickHouse if you need it one day. 
As for me, I need to make more efforts with my studies at the moment, so it is unlikely I will return to this pull request in the future.

**Comment by rschu1ze** [comment]
@aterehov Appreciated it, thanks for the update! As mentioned above, I don't see an urgent need to have such a function in ClickHouse at the moment. For that reason, I'll close the PR for now. If someone has a different opinion, please let me know, we can reopen it then.

**Comment by rschu1ze** [comment]
Lol, I just noticed the feature was motivated by #6498. Okay then, I'll continue/merge it if I find some time.

**Comment by rschu1ze** [comment]
We can continue this another time.

**Pull request closed without merging**